### PR TITLE
[Rshaman] Typescript

### PR DESCRIPTION
**Pull request by niseko**
Converts most of Resto Shaman. Leftovers are difficult files that are going to be done 1 by 1, or files that will get deleted soon.

### CODE REVIEW

File: src/parser/shaman/restoration/modules/talents/Downpour.tsx [hunk 1]
```diff
@@ -35,15 +36,19 @@ class Downpour extends Analyzer {
   downpourHitsSum = 0;
   downpourTimestamp = 0;
 
-  constructor(...args) {
-    super(...args);
+  protected cooldownThroughputTracker!: CooldownThroughputTracker;
```
Comment (Dambroda) [inline src/parser/shaman/restoration/modules/talents/Downpour.tsx:39]: Same comment as above, would be best if these were right below the deps declaration I think

**Review by Dambroda — APPROVED** [review]
I can't leave a comment on the file because there was no +/-, but in CONFIG.tsx you can add typechecking by:
```typescript
import Config from 'parser/Config';

const CONFIG: Config = {...}

export default CONFIG;

```
Very well converted 👍 Approved, every comment is optional

File: src/parser/shaman/restoration/modules/core/HealingRainLocation.tsx [hunk 2]
```diff
@@ -2,40 +2,54 @@ import SPELLS from 'common/SPELLS';
...
+
+
 class HealingRainLocation extends Analyzer {
   static dependencies = {
     combatants: Combatants,
   };
   healingRainDiameter = 2100; // 5% margin of error
 
-  healingRainEvents = [];
+  protected combatants!: Combatants;
```
[the remark this answers is not in the document]
Comment (niseko) [reply src/parser/shaman/restoration/modules/core/HealingRainLocation.tsx:30]: yeah, that was my idea and must have forgot about it for a couple files

File: src/parser/shaman/restoration/modules/core/HealingRainLocation.tsx [hunk 3]
```diff
@@ -85,17 +99,17 @@ class HealingRainLocation extends Analyzer {
     return this.sumHealing(filteredEvents, healIncrease, healingRainLocation);
   }
 
-  sumHealing(eventsDuringRain, healIncrease, healingRainLocation) {
+  sumHealing(eventsDuringRain: Array<Event<any>>, healIncrease: number, healingRainLocation: Location) {
```
[the remark this answers is not in the document]
Comment (niseko) [reply src/parser/shaman/restoration/modules/core/HealingRainLocation.tsx]: thanks, not sure why I was expecting more than healevents

File: src/parser/shaman/restoration/modules/Abilities.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import SPELLS from 'common/SPELLS';
 import ISSUE_IMPORTANCE from 'parser/core/ISSUE_IMPORTANCE';
 import CoreAbilities from 'parser/core/modules/Abilities';
+import { AbilityTrackerAbility } from 'parser/core/modules/Ability';
 import { calculateCooldown } from '../../../shared/modules/spells/bfa/essences/VisionsOfPerfection';
 
 class Abilities extends CoreAbilities {
```
[the remark this answers is not in the document]
Comment (niseko) [reply src/parser/shaman/restoration/modules/Abilities.ts:7]: 👍

**Pull request merged**